{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "ddc9aa60-2a1d-423f-9a83-11a33b9b409f",
   "metadata": {},
   "source": [
    "### Deduplicate the UMI's\n",
    "Now we must dedup the umis. But prior to this step, there are a couple issues we must address."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "8111fdb8-e08c-4a4c-88b5-a1cd5e3e0ba0",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pysam\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "45bf70e4-98d9-4a6e-9065-0c06c6090055",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Geneid</th>\n",
       "      <th>Chr</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Length</th>\n",
       "      <th>outFiles/TLS23A_R1_MouseAligned.sortedByCoord.out.bam</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSMUSG00000102693.2</td>\n",
       "      <td>chr1</td>\n",
       "      <td>3143476</td>\n",
       "      <td>3144545</td>\n",
       "      <td>+</td>\n",
       "      <td>1070</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSMUSG00000064842.3</td>\n",
       "      <td>chr1</td>\n",
       "      <td>3172239</td>\n",
       "      <td>3172348</td>\n",
       "      <td>+</td>\n",
       "      <td>110</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSMUSG00000051951.6</td>\n",
       "      <td>chr1;chr1;chr1;chr1;chr1;chr1;chr1</td>\n",
       "      <td>3276124;3276746;3283662;3283832;3284705;349192...</td>\n",
       "      <td>3277540;3277540;3285855;3286567;3287191;349212...</td>\n",
       "      <td>-;-;-;-;-;-;-</td>\n",
       "      <td>6094</td>\n",
       "      <td>19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSMUSG00000102851.2</td>\n",
       "      <td>chr1</td>\n",
       "      <td>3322980</td>\n",
       "      <td>3323459</td>\n",
       "      <td>+</td>\n",
       "      <td>480</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSMUSG00000103377.2</td>\n",
       "      <td>chr1</td>\n",
       "      <td>3435954</td>\n",
       "      <td>3438772</td>\n",
       "      <td>-</td>\n",
       "      <td>2819</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55354</th>\n",
       "      <td>ENSMUSG00000064368.1</td>\n",
       "      <td>chrM</td>\n",
       "      <td>13552</td>\n",
       "      <td>14070</td>\n",
       "      <td>-</td>\n",
       "      <td>519</td>\n",
       "      <td>1191</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55355</th>\n",
       "      <td>ENSMUSG00000064369.1</td>\n",
       "      <td>chrM</td>\n",
       "      <td>14071</td>\n",
       "      <td>14139</td>\n",
       "      <td>-</td>\n",
       "      <td>69</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55356</th>\n",
       "      <td>ENSMUSG00000064370.1</td>\n",
       "      <td>chrM</td>\n",
       "      <td>14145</td>\n",
       "      <td>15288</td>\n",
       "      <td>+</td>\n",
       "      <td>1144</td>\n",
       "      <td>4359</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55357</th>\n",
       "      <td>ENSMUSG00000064371.1</td>\n",
       "      <td>chrM</td>\n",
       "      <td>15289</td>\n",
       "      <td>15355</td>\n",
       "      <td>+</td>\n",
       "      <td>67</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55358</th>\n",
       "      <td>ENSMUSG00000064372.1</td>\n",
       "      <td>chrM</td>\n",
       "      <td>15356</td>\n",
       "      <td>15422</td>\n",
       "      <td>-</td>\n",
       "      <td>67</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>55359 rows Ã— 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                     Geneid                                 Chr  \\\n",
       "0      ENSMUSG00000102693.2                                chr1   \n",
       "1      ENSMUSG00000064842.3                                chr1   \n",
       "2      ENSMUSG00000051951.6  chr1;chr1;chr1;chr1;chr1;chr1;chr1   \n",
       "3      ENSMUSG00000102851.2                                chr1   \n",
       "4      ENSMUSG00000103377.2                                chr1   \n",
       "...                     ...                                 ...   \n",
       "55354  ENSMUSG00000064368.1                                chrM   \n",
       "55355  ENSMUSG00000064369.1                                chrM   \n",
       "55356  ENSMUSG00000064370.1                                chrM   \n",
       "55357  ENSMUSG00000064371.1                                chrM   \n",
       "55358  ENSMUSG00000064372.1                                chrM   \n",
       "\n",
       "                                                   Start  \\\n",
       "0                                                3143476   \n",
       "1                                                3172239   \n",
       "2      3276124;3276746;3283662;3283832;3284705;349192...   \n",
       "3                                                3322980   \n",
       "4                                                3435954   \n",
       "...                                                  ...   \n",
       "55354                                              13552   \n",
       "55355                                              14071   \n",
       "55356                                              14145   \n",
       "55357                                              15289   \n",
       "55358                                              15356   \n",
       "\n",
       "                                                     End         Strand  \\\n",
       "0                                                3144545              +   \n",
       "1                                                3172348              +   \n",
       "2      3277540;3277540;3285855;3286567;3287191;349212...  -;-;-;-;-;-;-   \n",
       "3                                                3323459              +   \n",
       "4                                                3438772              -   \n",
       "...                                                  ...            ...   \n",
       "55354                                              14070              -   \n",
       "55355                                              14139              -   \n",
       "55356                                              15288              +   \n",
       "55357                                              15355              +   \n",
       "55358                                              15422              -   \n",
       "\n",
       "       Length  outFiles/TLS23A_R1_MouseAligned.sortedByCoord.out.bam  \n",
       "0        1070                                                  0      \n",
       "1         110                                                  0      \n",
       "2        6094                                                 19      \n",
       "3         480                                                  0      \n",
       "4        2819                                                  9      \n",
       "...       ...                                                ...      \n",
       "55354     519                                               1191      \n",
       "55355      69                                                  0      \n",
       "55356    1144                                               4359      \n",
       "55357      67                                                  0      \n",
       "55358      67                                                  0      \n",
       "\n",
       "[55359 rows x 7 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Lets take a look at some of the gene assigned counts which is the last column in this file\n",
    "gene_assigned = pd.read_csv('outFiles/LS23A_GeneAssigned', sep='\\t', header=1)\n",
    "gene_assigned"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "27d799b0-49b3-4beb-9649-da1b8c1733e5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Geneid</th>\n",
       "      <th>Chr</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Length</th>\n",
       "      <th>outFiles/TLS23A_R1_MouseAligned.sortedByCoord.out.bam</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>47428</th>\n",
       "      <td>ENSMUSG00000119584.1</td>\n",
       "      <td>chr17</td>\n",
       "      <td>40157244</td>\n",
       "      <td>40159092</td>\n",
       "      <td>+</td>\n",
       "      <td>1849</td>\n",
       "      <td>331191</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55323</th>\n",
       "      <td>ENSMUSG00000064337.1</td>\n",
       "      <td>chrM</td>\n",
       "      <td>70</td>\n",
       "      <td>1024</td>\n",
       "      <td>+</td>\n",
       "      <td>955</td>\n",
       "      <td>25327</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55325</th>\n",
       "      <td>ENSMUSG00000064339.1</td>\n",
       "      <td>chrM</td>\n",
       "      <td>1094</td>\n",
       "      <td>2675</td>\n",
       "      <td>+</td>\n",
       "      <td>1582</td>\n",
       "      <td>14313</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29423</th>\n",
       "      <td>ENSMUSG00000035202.9</td>\n",
       "      <td>chr9;chr9;chr9;chr9;chr9;chr9;chr9;chr9;chr9;c...</td>\n",
       "      <td>123195992;123195995;123196012;123196019;123196...</td>\n",
       "      <td>123196172;123197592;123196172;123196172;123196...</td>\n",
       "      <td>+;+;+;+;+;+;+;+;+;+;+;+;+;+;+;+;+;+;+;+;+;+;+;...</td>\n",
       "      <td>11089</td>\n",
       "      <td>9371</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55337</th>\n",
       "      <td>ENSMUSG00000064351.1</td>\n",
       "      <td>chrM</td>\n",
       "      <td>5328</td>\n",
       "      <td>6872</td>\n",
       "      <td>+</td>\n",
       "      <td>1545</td>\n",
       "      <td>5186</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21099</th>\n",
       "      <td>ENSMUSG00000090341.2</td>\n",
       "      <td>chr7</td>\n",
       "      <td>41222701</td>\n",
       "      <td>41223169</td>\n",
       "      <td>-</td>\n",
       "      <td>469</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21100</th>\n",
       "      <td>ENSMUSG00000108482.2</td>\n",
       "      <td>chr7;chr7;chr7</td>\n",
       "      <td>41244316;41244715;41245945</td>\n",
       "      <td>41244550;41244814;41246155</td>\n",
       "      <td>+;+;+</td>\n",
       "      <td>546</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21103</th>\n",
       "      <td>ENSMUSG00000091401.2</td>\n",
       "      <td>chr7</td>\n",
       "      <td>41315941</td>\n",
       "      <td>41317542</td>\n",
       "      <td>-</td>\n",
       "      <td>1602</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21105</th>\n",
       "      <td>ENSMUSG00000091004.3</td>\n",
       "      <td>chr7</td>\n",
       "      <td>41318032</td>\n",
       "      <td>41318731</td>\n",
       "      <td>-</td>\n",
       "      <td>700</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55358</th>\n",
       "      <td>ENSMUSG00000064372.1</td>\n",
       "      <td>chrM</td>\n",
       "      <td>15356</td>\n",
       "      <td>15422</td>\n",
       "      <td>-</td>\n",
       "      <td>67</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>55359 rows Ã— 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                     Geneid  \\\n",
       "47428  ENSMUSG00000119584.1   \n",
       "55323  ENSMUSG00000064337.1   \n",
       "55325  ENSMUSG00000064339.1   \n",
       "29423  ENSMUSG00000035202.9   \n",
       "55337  ENSMUSG00000064351.1   \n",
       "...                     ...   \n",
       "21099  ENSMUSG00000090341.2   \n",
       "21100  ENSMUSG00000108482.2   \n",
       "21103  ENSMUSG00000091401.2   \n",
       "21105  ENSMUSG00000091004.3   \n",
       "55358  ENSMUSG00000064372.1   \n",
       "\n",
       "                                                     Chr  \\\n",
       "47428                                              chr17   \n",
       "55323                                               chrM   \n",
       "55325                                               chrM   \n",
       "29423  chr9;chr9;chr9;chr9;chr9;chr9;chr9;chr9;chr9;c...   \n",
       "55337                                               chrM   \n",
       "...                                                  ...   \n",
       "21099                                               chr7   \n",
       "21100                                     chr7;chr7;chr7   \n",
       "21103                                               chr7   \n",
       "21105                                               chr7   \n",
       "55358                                               chrM   \n",
       "\n",
       "                                                   Start  \\\n",
       "47428                                           40157244   \n",
       "55323                                                 70   \n",
       "55325                                               1094   \n",
       "29423  123195992;123195995;123196012;123196019;123196...   \n",
       "55337                                               5328   \n",
       "...                                                  ...   \n",
       "21099                                           41222701   \n",
       "21100                         41244316;41244715;41245945   \n",
       "21103                                           41315941   \n",
       "21105                                           41318032   \n",
       "55358                                              15356   \n",
       "\n",
       "                                                     End  \\\n",
       "47428                                           40159092   \n",
       "55323                                               1024   \n",
       "55325                                               2675   \n",
       "29423  123196172;123197592;123196172;123196172;123196...   \n",
       "55337                                               6872   \n",
       "...                                                  ...   \n",
       "21099                                           41223169   \n",
       "21100                         41244550;41244814;41246155   \n",
       "21103                                           41317542   \n",
       "21105                                           41318731   \n",
       "55358                                              15422   \n",
       "\n",
       "                                                  Strand  Length  \\\n",
       "47428                                                  +    1849   \n",
       "55323                                                  +     955   \n",
       "55325                                                  +    1582   \n",
       "29423  +;+;+;+;+;+;+;+;+;+;+;+;+;+;+;+;+;+;+;+;+;+;+;...   11089   \n",
       "55337                                                  +    1545   \n",
       "...                                                  ...     ...   \n",
       "21099                                                  -     469   \n",
       "21100                                              +;+;+     546   \n",
       "21103                                                  -    1602   \n",
       "21105                                                  -     700   \n",
       "55358                                                  -      67   \n",
       "\n",
       "       outFiles/TLS23A_R1_MouseAligned.sortedByCoord.out.bam  \n",
       "47428                                             331191      \n",
       "55323                                              25327      \n",
       "55325                                              14313      \n",
       "29423                                               9371      \n",
       "55337                                               5186      \n",
       "...                                                  ...      \n",
       "21099                                                  0      \n",
       "21100                                                  0      \n",
       "21103                                                  0      \n",
       "21105                                                  0      \n",
       "55358                                                  0      \n",
       "\n",
       "[55359 rows x 7 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Sort by highest number of counts\n",
    "gene_assigned.sort_values('outFiles/TLS23A_R1_MouseAligned.sortedByCoord.out.bam', ascending=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "75bc3a12-3268-485f-a428-fa2a49650e1c",
   "metadata": {},
   "source": [
    "### Top two genes\n",
    "Note that the top two genes take up almost the majority of the reads. This can change depending on the GTF file you use.  \n",
    "For example we did not have this issue with an older GTF file.  \n",
    "These reads correspond to a Rn18s and mt-rRNA read. This also seems to be an issue specific to the primer we use in addition to these being highly abundant sequences.  \n",
    "\n",
    "For this sequencing run, the number of counts are not too bad, but for some of the higher sequencing depths, these two reads can take up over a million reads or so.  \n",
    "This is a problem for our UMI library as we use a 3 letter, 12 base UMI code.  \n",
    "As you can imagine, this only allows for $3^{12} = 531441$ maximum possibilities. Its actually a little bit less because the umi-tools also allows for a 1 base traversal difference when its deduplicating.  \n",
    "In our experience when we tried to deduplicate a read with this many counts, the deduplication pipeline stalled indefinitely, in fact it even stopped the o2 job because it was requesting inordinate amounts of memory (>64 gB).  \n",
    "\n",
    "To solve this problem, we actually removed these two reads and created a new BAM file prior to deduplication.  \n",
    "For your experimental system, this may or may not be a problem."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "a382179b-1d14-4d4a-a240-ede2bf4dc9e5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# This is the part of the script that removes the two counts\n",
    "from collections import defaultdict\n",
    "GENES_TO_FILTER = ['ENSMUSG00000119584.1', 'ENSMUSG00000064337.1']\n",
    "\n",
    "def createSmallerGeneFile(filePrefix, geneFile):\n",
    "    outGeneFile = '%s.bam' % filePrefix\n",
    "\n",
    "    gF = pysam.AlignmentFile(geneFile)\n",
    "    outGF = pysam.AlignmentFile(outGeneFile, 'w', template=gF)\n",
    "    \n",
    "    filteredCounts = defaultdict(int)\n",
    "\n",
    "    for read in gF.fetch(until_eof=True):\n",
    "        if read.has_tag('XT'):\n",
    "            xtTag = read.get_tag('XT')\n",
    "\n",
    "            if xtTag in GENES_TO_FILTER:\n",
    "                filteredCounts[xtTag] += 1\n",
    "            else:\n",
    "                outGF.write(read)\n",
    "\n",
    "    gF.close()\n",
    "    outGF.close()\n",
    "\n",
    "    print('Filtered out %s' % ', '.join(['%d of %s' % \\\n",
    "                                    (filteredCounts[gene], gene) \\\n",
    "                                    for gene in GENES_TO_FILTER]))\n",
    "\n",
    "    return outGeneFile"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "65c9bea5-7a66-47a0-a3aa-acb3e01675f6",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[E::idx_find_and_load] Could not retrieve index file for 'outFiles/TLS23A_R1_MouseAligned.sortedByCoord.out.bam.featureCounts.bam'\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtered out 331191 of ENSMUSG00000119584.1, 25327 of ENSMUSG00000064337.1\n"
     ]
    }
   ],
   "source": [
    "filePrefix = 'outFiles/LS23A_filtered'\n",
    "countedGeneFile = 'outFiles/TLS23A_R1_MouseAligned.sortedByCoord.out.bam.featureCounts.bam'\n",
    "filteredCountsFile = createSmallerGeneFile(filePrefix, countedGeneFile)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "77f83b58-5b8d-4cdf-9477-05b773afed36",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'outFiles/LS23A_filtered.bam'"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "filteredCountsFile"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "8b0b3176-7cb4-4e88-913d-61b9c736c54f",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Now we sort the file and index the file. UMI-tools requires this prior to doing the dedup function.\n",
    "!samtools sort outFiles/LS23A_filtered.bam -o outFiles/LS23A_Sorted.bam\n",
    "!samtools index outFiles/LS23A_Sorted.bam"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0fe90161-9285-4aa4-b688-4b89fe54127c",
   "metadata": {},
   "source": [
    "### Umi-tools deduplication\n",
    "For LS, we use the `--per-gene` deduplication option.  \n",
    "This means that if you have two reads, and one read maps to bases 1-20 for Gene A, and the second read maps to bases 2-21,  \n",
    "umi-tools will still consider these to be the same UMI group for the purposes of deduplication.  \n",
    "This option is recommended if there is a fragmentation step in your library prep, which there usually is.  \n",
    "\n",
    "For LS we do do a fragmentation step so we have this option on.  \n",
    "Previously we deduped without this option so it will dedup by mapped position instead of by gene.  \n",
    "It ultimately did not make a huge difference for us but its better to keep this option on.  \n",
    "\n",
    "Note that we must enter in the  featurecounts tag notation when using the `--per-gene` option, `XT` and `XS`.  \n",
    "If you use another package to assign gene counts, you must change these tags to the ones the other package uses."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "a0c1d688-2d66-4920-ace7-7484a3c57d3b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 101 ms, sys: 34.8 ms, total: 136 ms\n",
      "Wall time: 9.59 s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "# And finally deduplicate\n",
    "!umi_tools dedup --per-gene --gene-tag=XT --assigned-status-tag=XS -I outFiles/LS23A_Sorted.bam -S outFiles/LS23A_Dedup.bam -L outFiles/LS23A_dedup.log "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5c570d6e-10dc-42a1-81bf-1aa4f9c89bf6",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "LSEnv",
   "language": "python",
   "name": "lsenv"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
