{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "4b09c520-2472-49f7-96f9-1a3c4e1ee87f",
   "metadata": {
    "tags": []
   },
   "source": [
    "### Running scripts on the O2 cluster\n",
    "Running scripts on the o2 cluster will require you to set up the \"LSEnv\" on your o2 account.  \n",
    "The first part is making sure conda is installed and running whenever you start a session on o2.\n",
    "\n",
    "First step is we log in:\n",
    "`ssh nl82@o2.hms.harvard.edu`  \n",
    "\n",
    "Which will land you on the login node.  \n",
    "`(base) nl82@login05:~$ `  \n",
    "\n",
    "On the o2 website, there is a set of instructions that tell you how to install conda on your acct.  \n",
    "https://harvardmed.atlassian.net/wiki/spaces/O2/pages/1594263516/Conda+on+O2  \n",
    "You can largely just follow those directions, and you should be good.  \n",
    "\n",
    "For myself, I would start an interactive compute session as conda installs can be quite computationally intensive and if you run them on the login node. it will kick you off.  \n",
    "`srun -p interactive -t 12:00:00 --mem=16G --pty bash`  \n",
    "\n",
    "Once you have an interactive session, if you type in `module avail`  \n",
    "There are two conda versions available:  \n",
    "```\n",
    "conda2/4.2.13            (E)  \n",
    "miniconda3/4.10.3          (E)  \n",
    "```\n",
    "As of writing this, both miniconda 4.10.3 and conda 4.2.13 are still considered \"experimental\" hence the (E) tag.  \n",
    "Both modules should be fine.  \n",
    "As for myself, i decided to install conda on my o2 account with the following:  \n",
    "```\n",
    "module load conda2/4.2.13\n",
    "conda init bash\n",
    "```\n",
    "Once that has been set up, then close the terminal, restart and re-login to o2 and the `(base)` conda environment should automatically be activated whenever you log in to o2."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8b2cc216-3ead-463a-b609-0d55045c418b",
   "metadata": {},
   "source": [
    "### About the (base) conda environment on o2\n",
    "As far as i can tell, the base environment from the conda/4.2.13 module on o2 is not your typical base environment that you might have on your laptop.  \n",
    "One thing thats different is that the conda on o2 already has \"samtools\" installed in the base environment.  It is a common tool that alot of people use.  \n",
    "If you check the samtools path and version:  \n",
    "```\n",
    "(base) nl82@compute-a-16-167:~$ which samtools\n",
    "/n/app/conda2/bin/samtools\n",
    "\n",
    "(base) nl82@compute-a-16-167:~$ samtools --version\n",
    "samtools 1.9\n",
    "Using htslib 1.9\n",
    "Copyright (C) 2018 Genome Research Ltd.\n",
    "```\n",
    "You can see that the base conda environment already has samtools v1.9.  This is good to be aware of because it might lead to some confusing discrepancies.  \n",
    "For example if you run a python script that uses samtools on the o2 base conda environment, it will work fine. But it wont work if you run it in your (base) conda environment on your laptop.  "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "70d56274-d134-44d7-a873-9d0149913ba3",
   "metadata": {},
   "source": [
    "### Installing STAR aligner\n",
    "There are a couple options for using STAR on o2.  \n",
    "\n",
    "#### First Option\n",
    "If you check the available modules, you can see that we already have a few versions of STAR aligner on o2.  \n",
    "```\n",
    "star/2.7.9a                (D)\n",
    "```\n",
    "When you start up a session on o2, the STAR aligner is not loaded by default, so if you try to call the STAR package:  \n",
    "```\n",
    "(base) nl82@compute-a-16-167:~$ STAR\n",
    "bash: STAR: command not found\n",
    "```\n",
    "It will tell you that the command is not found.  \n",
    "\n",
    "You can load the star module like so (Note: you have to load the gcc module as well):\n",
    "```\n",
    "module load gcc/6.2.0\n",
    "module load star/2.7.9a\n",
    "\n",
    "Currently Loaded Modules:\n",
    "  1) gcc/6.2.0   2) star/2.7.9a\n",
    "```\n",
    "And if you run STAR: \n",
    "```\n",
    "(base) nl82@compute-a-16-167:/n/data1/hms/sysbio/yin/Ninning$ STAR\n",
    "Usage: STAR  [options]... --genomeDir /path/to/genome/index/   --readFilesIn R1.fq R2.fq\n",
    "Spliced Transcripts Alignment to a Reference (c) Alexander Dobin, 2009-2020\n",
    "\n",
    "STAR version=2.7.9a\n",
    "STAR compilation time,server,dir=2021-05-04T09:43:56-0400 vega:/home/dobin/data/STAR/STARcode/STAR.master/source\n",
    "For more details see:\n",
    "<https://github.com/alexdobin/STAR>\n",
    "<https://github.com/alexdobin/STAR/blob/master/doc/STARmanual.pdf>\n",
    "\n",
    "To list all parameters, run STAR --help\n",
    "```\n",
    "Then it will recognize the command.\n",
    "\n",
    "#### Second (preferred?) option\n",
    "Setting up STAR the first way will require you to load the modules everytime you start up o2. Alternatively you can install STAR into your folder and have your python scripts automatically call them when running the script, which is what we do for Light-Seq.  \n",
    "From the STAR aligner github, https://github.com/alexdobin/STAR  \n",
    "you can just follow the instructions for the linux install (o2 is running linux kernel 3.10.0-1160.45.1.el7.x86_64).  \n",
    "\n",
    "First, you would just copy the entire folder in your account, you can either upload the files through Filezilla,  \n",
    "Or while on o2, type:  \n",
    "```\n",
    "wget https://github.com/alexdobin/STAR/archive/2.7.10a.tar.gz\n",
    "tar -xzf 2.7.10a.tar.gz\n",
    "cd STAR-2.7.10a\n",
    "(base) nl82@compute-a-16-167:/n/data1/hms/sysbio/yin/Ninning/STAR-2.7.10a$ \n",
    "```\n",
    "And go to the 'source' directory, which should have the `STAR.cpp` c file which you would compile to install STAR in your account.  \n",
    "You still have to use a C compiler (based on my old notes), so make sure you load the c compiler before you compile the STAR file.  \n",
    "The default gcc for o2 when you first login is `gcc (GCC) 4.8.5`, when I last tried to compile STAR with this gcc version, it gave an endless array of warning messages and then stalled, even though on re-login it seemed to have worked. To save yourself trouble down the line, I would absolutely load the recommended gcc version on o2 before compiling.\n",
    "```\n",
    "cd source\n",
    "module load gcc/6.2.0\n",
    "make STAR\n",
    "```\n",
    "Once its compiled, you would call STAR from your bash terminal by directing it to the path like this `./STAR`, instead of calling it with just `STAR`.  I explain this further in another notebook.  \n",
    "At this point you can call the star aligner:  \n",
    "```\n",
    "(base) nl82@compute-e-16-229:/n/data1/hms/sysbio/yin/Ninning/STAR-2.7.10a/source$ ./STAR\n",
    "Usage: STAR  [options]... --genomeDir /path/to/genome/index/   --readFilesIn R1.fq R2.fq\n",
    "Spliced Transcripts Alignment to a Reference (c) Alexander Dobin, 2009-2020\n",
    "\n",
    "STAR version=2.7.10a\n",
    "STAR compilation time,server,dir=2022-04-27T18:00:41-0400 login06.o2.rc.hms.harvard.edu:/n/data1/hms/sysbio/yin/Ninning/STAR-2.7.10a/source\n",
    "For more details see:\n",
    "<https://github.com/alexdobin/STAR>\n",
    "<https://github.com/alexdobin/STAR/blob/master/doc/STARmanual.pdf>\n",
    "\n",
    "To list all parameters, run STAR --help\n",
    "```\n",
    "And it should be ready to use.  \n",
    "\n",
    "Final note: The permissions for the STAR executable should all be enabled `-rwxrwxr-x 1 nl82 yin 3192704 Apr 27 18:03 STAR`.  \n",
    "But sometimes its not. I havent encountered this random issue recently on o2 but its always good to check with `ls -l`.  \n",
    "If the permissions are off then you will have to enable them with the `chmod u+x STAR`."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "02d7ca39-dcb0-4fd0-b6c6-f81c561b8ea5",
   "metadata": {},
   "source": [
    "### Setting up the LSEnv on o2\n",
    "Once you have the base conda environment installed and STAR installed, setting up the \"LSEnv\" should not be that different than the instructions currently written for the Light-Seq github.  \n",
    "For example i will create a second LSEnv, (yours will just be `LSEnv`):  \n",
    "```\n",
    "(base) nl82@compute-e-16-229:/n/data1/hms/sysbio/yin/Ninning$ conda create --name LSEnv2 python=3.7.5 pandas matplotlib seaborn PyTables Biopython scikit-image\n",
    "(base) nl82@compute-e-16-229:/n/data1/hms/sysbio/yin/Ninning$ conda activate LSEnv2\n",
    "\n",
    "(LSEnv2) nl82@compute-e-16-229:/n/data1/hms/sysbio/yin/Ninning$ python --version\n",
    "Python 3.7.5\n",
    "```\n",
    "At which point you can activate your LSEnv version, check the python version to make sure its installed properly, and proceed with the rest of the install.  "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6bb3ff12-c74d-487c-9220-fb32eef36204",
   "metadata": {},
   "source": [
    "### Submitting a batch job on o2\n",
    "Typically you would submit a batch job and forget about it and come back once the computation is done.  \n",
    "The o2 website has instructions on how to submit a batch job.  \n",
    "\n",
    "Essentially you would just create a `.sh` file, for example `job.sh`, with the following:  \n",
    "```\n",
    "#!/bin/bash\n",
    "#SBATCH -c 12\n",
    "#SBATCH -t 0-12:00    \n",
    "#SBATCH -p short\n",
    "#SBATCH --mem=64G\n",
    "#SBATCH -o 220415map_%j.out\n",
    "#SBATCH -e 220415map_%j.err\n",
    "\n",
    "module load conda2/4.2.13\n",
    "source activate LSEnv\n",
    "python3 MapToMouseUniqueOnly.py\n",
    "```\n",
    "And submit by typing:  \n",
    "```\n",
    "sbatch job.sh\n",
    "```\n",
    "\n",
    "In this file, we would set the number of cores we want (12), the time, partition etc... and also set the file names for the logs `220415map_%j.out`, which will include the date at the time of running the script and the job number. You can also set it to save the logs to another folder like `bashOut/220415map_%j.out` if you want to organize it.  \n",
    "For the commands we run, we still go through the motions of loading the conda module (in case its not automatically loaded), activating the LSEnv, then running the script with the python command.  \n",
    "In this case we used the `python3` command just to make sure we are using the correct python version.  \n",
    "If you are in the LSEnv, both `python` and `python3` should point to the same python version since thats the only version installed in \"LSEnv\".  \n",
    "```\n",
    "(LSEnv) nl82@compute-e-16-229:/n/data1/hms/sysbio/yin/Ninning$ python3 --version\n",
    "Python 3.7.5\n",
    "(LSEnv) nl82@compute-e-16-229:/n/data1/hms/sysbio/yin/Ninning$ python --version\n",
    "Python 3.7.5\n",
    "```\n",
    "However if you did something weird with your `python` path pointing on your o2 account prior to this, then you should double check to make sure that when you call the `python` or `python3` command, its actually calling the right one."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "14c7079b-9974-42c6-a476-87a7d6ecd0c3",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
